### PR TITLE
change ZKP section

### PR DESCRIPTION
**Pull request by brentzundel**
This PR:

- Adds a version of the note initially drafter by Orie in #1026 
- modernizes and generalizes the language in the section in an attempt to make it more applicable to various zero knowledge proof mechanisms.
- adds a note that the examples will either be re-worked or removed.

I believe this PR prepares this section for for one of these possibilities:
1. The WG normatively specifies a method for securing VCs that supports ZKPs, in which case the examples and normative language in this section could be easily updated.
2. The WG does not normatively specify a method for securing VCs that supports ZKPs, in which case the examples and normative language in this section would be removed, and the remaining language could be moved to the implementation guide.

Signed-off-by: Brent Zundel <brent.zundel@gmail.com>

***
<a href="https://pr-preview.s3.amazonaws.com/brentzundel/vc-data-model/pull/1030.html" title="Last updated on Mar 21, 2023, 6:46 PM UTC (5cbc6a4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/1030/aeaa1ce...brentzundel:5cbc6a4.html" title="Last updated on Mar 21, 2023, 6:46 PM UTC (5cbc6a4)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -2907,63 +2907,66 @@ <h3>Evidence</h3>
...
-identifiable information contained on the degree.
+Zero-knowledge proofs are cryptographic methods which enable a user to prove
+knowledge of a value without disclosing the actual value. This data model
+supports being secured with the use of zero-knowledge proof mechanisms.
         </p>
         <p>
-The key capabilities introduced by zero-knowledge proof mechanisms are the
-ability of a <a>holder</a> to:
+Some capabilities that are compatible with <a>verifiable credentials</a> that
+made possible by zero-knowledge proof mechanisms:
```
Comment (TallTed) [inline index.html:2924]:
```suggestion
The following are some capabilities that are compatible with
<a>verifiable credentials</a>, that are made possible by zero-knowledge
proof mechanisms:
```

File: index.html [hunk 2]
```diff
@@ -2907,63 +2907,66 @@ <h3>Evidence</h3>
...
+involve the <a>issuer</a> during presentation. This provides a great deal of
+flexibility for <a>holders</a> to use their issued
+<a>verifiable credentials</a>.
           </li>
           <li>
-Selectively disclose the <a>claims</a> in a <a>verifiable credential</a> to a
-<a>verifier</a> without requiring the issuance of multiple atomic
-<a>verifiable credentials</a>. This allows a <a>holder</a> to provide a
-<a>verifier</a> with precisely the information they need and nothing more.
+Blinding of the signature value that is shared with a <a>verifier</a>. This
```
Comment (TallTed) [inline index.html:2938]:
```suggestion
<b>Blinding of the signature value that is shared with a <a>verifier</a>.</b> This
```

File: index.html [hunk 3]
```diff
@@ -2907,63 +2907,66 @@ <h3>Evidence</h3>
...
+allows a <a>holder</a> to share a different signature value with each
+presentation, which in turn reduces the amount of data shared.
           </li>
           <li>
-Produce a derived <a>verifiable credential</a> that is formatted according to
-the <a>verifier's</a> data schema instead of the <a>issuer's</a>, without
-needing to involve the <a>issuer</a> after <a>verifiable credential</a>
-issuance. This provides a great deal of flexibility for <a>holders</a> to use
-their issued <a>verifiable credentials</a>.
+Privacy preserving identification of the <a>holder</a>/<a>subject</a>. This
```
Comment (TallTed) [inline index.html:2943]:
```suggestion
<b>Privacy preserving identification of the <a>holder</a> and/or <a>subject</a>.</b> This
```

File: index.html [hunk 4]
```diff
@@ -2907,63 +2907,66 @@ <h3>Evidence</h3>
...
-how the data model can be used to issue, present, and verify zero-knowledge
-<a>verifiable credentials</a>.
+Not all capabilities are supported in all zero-knowledge proof mechanisms.
+Specific details about the capabilities and techniques provided by a particular
+zero knowledge proof mechanism, along with any normative requirements for using
+them with <a>verifiable credentials</a>, would be found in a specification for
+securing <a>verifiable credentials</a> with that zero-knowledge proof mechanism.
+        </p>
+        <p>
+We note that in most instances, for <a>holder</a> to make use of zero knowledge
```
Comment (TallTed) [inline index.html:2961]:
```suggestion
We note that in most instances, for a <a>holder</a> to make use of zero knowledge
```

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-02-08-vcwg#section3-2) on 2023-02-08

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 3.2. change ZKP section (pr vc-data-model#1030)

_See github pull request [vc-data-model#1030](https://github.com/w3c/vc-data-model/pull/1030)._




**Brent Zundel:** should be straightforward to adjust further or get rid of the section.  

**Manu Sporny:** orie suggests process in order. Doing that.  
… processing #1026, #1028 then need review on #1030.  

</details>

File: index.html [hunk 5]
```diff
@@ -2907,63 +2907,67 @@ <h3>Evidence</h3>
 
       <section>
         <h3>Zero-Knowledge Proofs</h3>
-
+        <p class="issue" data-number="863">
+We plan to remove the normative requirements in this section if there are no
+active work items for securing <a>verifiable credentials</a> using zero
+knowledge proof mechanisms when we transition v2 of the data model to CR.
```
Comment (quartzjer) [inline index.html:2913]:
```suggestion
knowledge proof mechanisms when we transition to v2 of the data model to CR.
```

File: index.html [hunk 6]
```diff
@@ -2978,7 +2982,15 @@ <h3>Zero-Knowledge Proofs</h3>
...
           </li>
         </ul>
-
+        <p>
+The examples below highlight how the data model might be used to issue and
+present <a>verifiable credentials</a> in zero-knowledge.
+        </p>
+        <p class="note">
+The provided examples will either be significantly re-written to demonstrate how
+to secure a <a>verifiable credential</a> using a normatively defined method that
```
Comment (quartzjer) [inline index.html:2991]:
```suggestion
to secure a <a>verifiable credential</a> using normatively defined methods that
```

**Review by quartzjer — APPROVED** [review]

File: index.html [hunk 7]
```diff
@@ -2907,63 +2907,67 @@ <h3>Evidence</h3>
...
+Zero-knowledge proofs are cryptographic methods which enable a user to prove
+knowledge of a value without disclosing the actual value. This data model
+supports being secured with the use of zero-knowledge proof mechanisms.
         </p>
         <p>
-The key capabilities introduced by zero-knowledge proof mechanisms are the
-ability of a <a>holder</a> to:
+The following are some capabilities that are compatible with
+<a>verifiable credentials</a>, that are made possible by zero-knowledge
+proof mechanisms:
```
Comment (dlongley) [inline index.html:2925]:
```suggestion
<a>verifiable credentials</a>, made possible by zero-knowledge
proof mechanisms:
```

File: index.html [hunk 8]
```diff
@@ -2907,63 +2907,67 @@ <h3>Evidence</h3>
...
         <ul>
           <li>
-Combine multiple <a>verifiable credentials</a> from multiple <a>issuers</a> into
-a single <a>verifiable presentation</a> without revealing
-<a>verifiable credential</a> or <a>subject</a> identifiers to the
-<a>verifier</a>. This makes it more difficult for the <a>verifier</a> to collude
-with any of the issuers regarding the issued <a>verifiable credentials</a>.
+<b>Selective disclosure of the properties in a <a>verifiable credential</a> by the
+<a>holder</a> to a <a>verifier</a>.</b> This allows a <a>holder</a> to provide a
+<a>verifier</a> with precisely the information they need and nothing more.
```
Comment (dlongley) [inline index.html:2931]:
```suggestion
<a>verifier</a> with precisely the information they request and nothing more.
```

**Review by dlongley — APPROVED** [review]
I remain a bit concerned about over promising what ZKP can realistically do in various use cases, but this text hasn't changed that much from what was originally approved to be in the VCDM 1.1 spec and there will be another revision at some point in the future before 2.0 is finished, so +1. Thanks for cleaning this up!

**Review by Sakurann — COMMENTED** [review]
Selective disclosure, derived claim and unlinkability are three distinct features. By ZKP do we mean a way to secure VCs that fulfill all three or at least one? This needs to be clear. In the former, I think it is highly unlikely we will have a mature way before VC-data-model v2 CR, if the latter, SD-JWT will fall under this umbrella, but then the section name should be renamed from "ZKP".

**Comment by brentzundel** [comment]
> Selective disclosure, derived claim and unlinkability are three distinct features. By ZKP do we mean a way to secure VCs that fulfill all three or at least one? This needs to be clear. In the former, I think it is highly unlikely we will have a mature way before VC-data-model v2 CR, if the latter, SD-JWT will fall under this umbrella, but then the section name should be renamed from "ZKP".

I agree that if SD-JWT becomes a normative part of the spec, then this section should be expanded to include it.

File: index.html [hunk 9]
```diff
@@ -2908,9 +2908,10 @@ <h3>Evidence</h3>
       <section>
         <h3>Zero-Knowledge Proofs</h3>
         <p class="issue" data-number="863">
-We plan to remove the normative requirements in this section if there are no
+We anticipate this section will undergo major revision for v2 of the data model,
+and plan to remove the normative requirements in this section if there are no
 active work items for securing <a>verifiable credentials</a> using zero
-knowledge proof mechanisms when we transition v2 of the data model to CR.
+knowledge proof mechanisms when we transition to v2 of the data model to CR.
```
Comment (TallTed) [inline index.html:2914]: If this doesn't make sense to you, try reading both versions of the line aloud.
```suggestion
knowledge proof mechanisms when we transition v2 of the data model to CR.
```

File: index.html [hunk 10]
```diff
@@ -2958,7 +2960,7 @@ <h3>Zero-Knowledge Proofs</h3>
 securing <a>verifiable credentials</a> with that zero-knowledge proof mechanism.
         </p>
         <p>
-We note that in most instances, for <a>holder</a> to make use of zero knowledge
+We note that in most instances, for a <a>holder</a> to make use of zero knowledge
 mechanisms with <a>verifiable credentials</a> requires an <a>issuer</a> to
 secure the <a>verifiable credential</a> in a manner that supports these
```
Comment (TallTed) [inline index.html:2965]:
```suggestion
mechanisms with <a>verifiable credentials</a> requires the <a>issuer</a> to
have secured the <a>verifiable credential</a> in a manner that supports these
```

**Review by TallTed — CHANGES_REQUESTED** [review]

File: index.html [hunk 11]
```diff
@@ -2907,70 +2907,68 @@ <h3>Evidence</h3>
 
       <section>
         <h3>Zero-Knowledge Proofs</h3>
```
Comment (Sakurann) [inline index.html:2909]:
```suggestion
        <h3>Selective Disclosure and Unlinkability</h3>
```

**Review by Sakurann — CHANGES_REQUESTED** [review]
it should be clear that the properties being discussed are mainly Selective Disclosure and unlinkability (via derivatives and identifier clinding), and one mechanism does not have to achieve all three at the same time.

File: index.html [hunk 12]
```diff
@@ -2985,7 +2983,15 @@ <h3>Zero-Knowledge Proofs</h3>
 operations in zero-knowledge.
           </li>
         </ul>
-
+        <p>
+The examples below highlight how the data model might be used to issue and
+present <a>verifiable credentials</a> in zero-knowledge.
+        </p>
+        <p class="note">
```
Comment (msporny) [inline index.html:2990]:
```suggestion
        <p class="issue" title="ZKP examples need to be updated">
```

**Review by msporny — APPROVED** [review]
Approving as long as minor nits are fixed.

File: index.html [hunk 13]
```diff
@@ -2906,71 +2906,69 @@ <h3>Evidence</h3>
...
+zero knowledge proof mechanism, along with any normative requirements for using
+them with <a>verifiable credentials</a>, would be found in a specification for
+securing <a>verifiable credentials</a> with that zero-knowledge proof mechanism.
+        </p>
+        <p>
+We note that in most instances, for a <a>holder</a> to make use of zero knowledge
+mechanisms with <a>verifiable credentials</a> requires the <a>issuer</a> to
+have secured the <a>verifiable credential</a> in a manner that supports these
+capabilities.
         </p>
```
Comment (brentzundel) [inline index.html:2967]:
```suggestion
capabilities.
        </p>
        <p>
It should also be noted that zero-knowledge proof mechanisms are not the only
means of providing the above-mentioned capabilities. 
        </p>
```

**Comment by brentzundel** [comment]
@Sakurann and @TallTed I have made changes to the PR to reflect your comments. Please re-review.

**Review by TallTed — APPROVED** [review]
Much better than it was.

**Comment by msporny** [comment]
Merging this is currently blocked by @Sakurann. I've requested an explicit re-review. @Sakurann, please re-review the changes that @brentzundel has made to this PR so we can either merge it, or make further changes to address your concerns.

**Review by tplooker — APPROVED** [review]

File: index.html [hunk 14]
```diff
@@ -2906,71 +2906,73 @@ <h3>Evidence</h3>
...
+<a>subject</a> identifiers to the <a>verifier</a>.
           </li>
         </ul>
-
         <p>
-This specification describes a data model that supports selective disclosure
-with the use of zero-knowledge proof mechanisms. The examples below highlight
-how the data model can be used to issue, present, and verify zero-knowledge
-<a>verifiable credentials</a>.
+Not all capabilities are supported in all zero-knowledge proof mechanisms.
```
Comment (Sakurann) [inline index.html:2956]:
```suggestion
Not all capabilities are supported in each zero-knowledge proof mechanisms.
```

**Review by Sakurann — APPROVED** [review]
Despite the change in the title, the section is still all about "zero-knowledge proof", which is a specific term and would not include things like SD-JWT. But given there are caveats to rewrite/remove this section written in the text, will keep the reservations until when that PR comes in later.

File: index.html (hunk 14 above)
Comment (TallTed) [reply index.html:2956]:
```suggestion
A given zero-knowledge proof mechanism might not support all capabilities described here.
```

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-03-14-vcwg-special#section2-2) on 2023-03-14

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.2. change ZKP section (pr vc-data-model#1030)

_See github pull request [vc-data-model#1030](https://github.com/w3c/vc-data-model/pull/1030)._




**Manu Sporny:** I think we're ready to merge this.  
… There are some changes that Kristina and Ted have suggested, but I think those are good to go unless someone here objects.  

> *Dave Longley:* +1 to merge with whatever editorial fixes.

**Brent Zundel:** ok, I'll apply language and merge.  

</details>

File: index.html [hunk 15]
```diff
@@ -2908,69 +2908,66 @@ <h3>Evidence</h3>
...
-identifiable information contained on the degree.
+Zero-knowledge proofs are cryptographic methods which enable a user to prove
+knowledge of a value without disclosing the actual value. This data model
+supports being secured with the use of zero-knowledge proof mechanisms.
         </p>
         <p>
-The key capabilities introduced by zero-knowledge proof mechanisms are the
-ability of a <a>holder</a> to:
+Some capabilities that are compatible with <a>verifiable credentials</a> that
+made possible by zero-knowledge proof mechanisms:
```
Comment (TallTed) [inline index.html:2925]:
```suggestion
Some capabilities that are compatible with <a>verifiable credentials</a> which
are made possible by zero-knowledge proof mechanisms:
```

File: index.html [hunk 16]
```diff
@@ -2908,69 +2908,66 @@ <h3>Evidence</h3>
...
-Produce a derived <a>verifiable credential</a> that is formatted according to
-the <a>verifier's</a> data schema instead of the <a>issuer's</a>, without
-needing to involve the <a>issuer</a> after <a>verifiable credential</a>
-issuance. This provides a great deal of flexibility for <a>holders</a> to use
-their issued <a>verifiable credentials</a>.
+Privacy preserving identification of the <a>holder</a>/<a>subject</a>. This
+allows a <a>holder</a> to prove that a <a>credential</a> was issued to them, or
+a <a>subject</a> to prove that a <a>credential</a> was issued about them,
+without sharing an identifier. This also reduces the amount of data necessary to
+share. This capability can also be used to combine multiple
```
Comment (TallTed) [inline index.html:2948]:
```suggestion
be shared. This capability can also be used to combine multiple
```

**Comment by David-Chadwick** [comment]
Have you considered the impact of these edits on section 5.4 (data schemas) as this has an example that concerns ZKPs. Are edits also needed to this section?

**Comment by Sebastian-Elfors-IDnow** [comment]
Thanks for the re-write of the ZKP section. It's a great improvement.

For completeness, the term "unlinkability" could be introduced. It is related to the bullet about "blinded signatures", which could be extended as follows: 

"Blinded signatures also allows for unlinkable proofs, which removes a common source of correlatable properties about a holder during multiple presentations to one or more verifiers."

**Pull request merged**